Automating GNS3 lab setup via Ansible

francescocappa opened the pull request:
# Description

This pull request introduces Ansible automation to streamline the GNS3 setup process. It replaces manual operations with automated tasks, making the setup more efficient and user-friendly.

# Steps

1. Download the Cisco IOS Image
2.  Start GNS3
3.  Create a new template for the Cisco 2691 router
4. Configure the Xfce4 terminal in GNS3
5. Add the network-multitool docker image to the list of end devices

Step 2 is typically performed using the GUI. To enhance the process and ensure proper systemd integration, we have created a custom gns3.service file.

# How Has This Been Tested?

This setup has been tested locally on the target host.

frisso commented on provisioning/virtual-machines/ansible/roles/gns3/tasks/main.yml:
```diff
@@ -75,3 +75,108 @@
     group: "{{ ansible_user }}"
     mode: '0644'
+
+### The following tasks are due to the previous manual setup of GNS3 environment
```
```suggestion
### The following Ansible tasks replace the previous manual setup of the GNS3 environment
```

frisso commented on provisioning/virtual-machines/ansible/roles/gns3/tasks/main.yml:
```diff
@@ -75,3 +75,108 @@
+
+## 2.  Create a new template for the Cisco 2691 router
+
+# GNS3 server is usually launched though UI. A terminal launch is needed.
```
```suggestion
# GNS3 server is usually launched through UI. Here we need to start it via command line.
```

frisso commented on provisioning/virtual-machines/ansible/roles/gns3/tasks/main.yml:
```diff
@@ -75,3 +75,108 @@
+# 2.3 The GNS3 server needs some seconds to be ready
+- name: Introduce a sleep delay of 10 seconds
+  pause:
+    seconds: 5
```
5 or 10 seconds?

QcFe replied:
As you prefer, in any case though, I see nothing wrong with waiting 5 more seconds, if it means to be safer and avoid wasting maybe 15 minutes in rebuilding from scratch :)

francescocappa replied:
The process should ideally take just a couple of seconds. Striking a balance between security and responsiveness, a 5-second timeframe would be appropriate.

QcFe commented:
Thanks for the PR!
Only thing, I'd use NFS for retrieving the needed files, as has been done here:
https://github.com/netgroup-polito/CrownLabs/commit/be3ac2a7f97a97d8c55dfb563e679767a5648573
(search for matlab_nfs_source to see how it has been done).

I already created a volume and added the image inside, I can send you privately the details for testing.

francescocappa commented:
@QcFe 
I have implemented the modifications as per your suggested. I introduced a variable called `gns3_device_images`, which points to an NFS path containing the required files. After testing it in a local environment, it appears to be functioning correctly. 
If you have any further feedback, please let me know

QcFe commented on provisioning/virtual-machines/ansible/roles/gns3/tasks/main.yml:
```diff
@@ -75,3 +75,114 @@
+    state: mounted
+
+# 1.3 Decompress the file from .bin to .image
+- name: Unzip the file
+  shell: >
+        unzip -p /mnt/gns3_images/c2691-advipservicesk9-mz.124-25c.bin >
+        /home/{{ ansible_user }}/GNS3/images/IOS/c2691-advipservicesk9-mz.124-25c.image
+  ignore_errors: yes
```
It's always better to use native Ansible directives when possible, to avoid incompatibilities and possibly easing results.
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/unarchive_module.html

francescocappa replied:
Yes. I did modify the tasks according to what you suggested.
Thanks!

QcFe commented on provisioning/virtual-machines/ansible/roles/gns3/tasks/main.yml:
```diff
@@ -75,3 +75,114 @@
+  ignore_errors: yes
+
+ # 1.4 Unmount GNS3 NFS install volume
+- name: Unmount Matlab NFS install volume
+  ansible.posix.mount:
+    src: "{{ gns3_device_images }}"
+    path: "/mnt/gns3_images"
+    opts: rw,sync,hard
+    boot: false
+    fstype: nfs
+    state: absent
```
is this duplicate?

QcFe commented on provisioning/virtual-machines/ansible/roles/gns3/tasks/main.yml:
```diff
@@ -75,3 +75,114 @@
+  pause:
+    seconds: 5
+
+# 2.4 Create the request body to be used in Cisco C2961 template creation
+- name: Copy Cisco 2961 template request body
+  copy:
+    src: "{{ role_path }}/files/cisco-c2961.json"
+    dest: /tmp/cisco-c2961.json
+    mode: '0644'
+
+# 2.5 Performing the request to the GNS3 server
+- name: Create new template using GNS3 API
+  command: sudo  curl -X POST "http://localhost:3080/v2/templates" -d @/tmp/cisco-c2961.json
+
+# 2.5 Remove the temporary file used to hold the request body
+- name: Remove JSON request file 
+  file:
+    path: /tmp/json_file.json
+    state: absent
```
You might probably manage to do this with a single shot with Ansible uri module
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/uri_module.html

francescocappa commented:
Done

QcFe commented:
/merge
